### PR TITLE
#22: Use a BFS Query-backed DB manager

### PR DESCRIPTION
**Pull request by JadedCtrl**
Uses a query DB manager as a drop-in replacement for the SQL manager. All SQL events are imported automatically, and behavior should be same-- with the exception of the 'default' category and 'All Day' events.

An option is added to preferences to change the default category, and any event starting at 00:00 and ending at 23:59 is counted as 'All Day' by the db manager (as opposed to using a dedicated trait).

This is my first go at anything really query-backed, so there's probably a good bit wrong here. But it seems like it's been working well on my systems for a few days.

### CODE REVIEW

File: src/MainWindow.cpp [hunk 1]
```diff
@@ -315,13 +311,15 @@ MainWindow::_SetEventListPopUpEnabled(bool state)
...
+	Preferences* preferences = ((App*)be_app)->GetPreferences();
+
+	if (preferences != NULL) {
+		if(preferences->fHeaderVisible == true)
 			fSidePanelView->ShowWeekHeader(true);
 		else
 			fSidePanelView->ShowWeekHeader(false);
 
-		fSidePanelView->SetStartOfWeek(fPreferences->fStartOfWeekOffset);
+		fSidePanelView->SetStartOfWeek(preferences->fStartOfWeekOffset);
```
Comment (pulkomandy) [inline src/MainWindow.cpp:322]: This gets me curious (but it's unrelated to this change): why is there a setting in the app for this? Shouldn't it use the Locale Kit to get this info from the OS preferences instead?

File: src/db/QueryDBManager.cpp [hunk 2]
```diff
@@ -0,0 +1,950 @@
...
+	volRoster.GetBootVolume(&fQueryVolume);
+
+	_CategoryMimetype();
+	_EventMimetype();
+	_AddIndices();
+
+	// Create default categories if need be
+	if (GetAllCategories()->CountItems() == 0) {
+		Category* defaultCategory =
+			new Category(B_TRANSLATE("Default"), BString("1E90FF"), "1f1e4ffd-527d-4796-953f-df2e2c600a09");
```
Comment (pulkomandy) [inline src/db/QueryDBManager.cpp:101]: what are these? UUIDs? Where do they come from?

File: src/db/QueryDBManager.cpp [hunk 3]
```diff
@@ -0,0 +1,950 @@
...
+
+	return _GetEventsOfInterval(weekStart, weekEnd);
+}
+
+
+BList*
+QueryDBManager::GetEventsOfCategory(Category* category)
+{
+	BList* events = new BList();
+	BQuery* query = new BQuery();
```
Comment (pulkomandy) [inline src/db/QueryDBManager.cpp:258]: why not simply `BQuery query;` ? (also in other functions below where you do the same)

File: src/db/QueryDBManager.cpp [hunk 4]
```diff
@@ -0,0 +1,950 @@
...
+
+	return _CategoryToFile(newCategory, catFile);
+}
+
+
+Category*
+QueryDBManager::GetCategory(const char* name)
+{
+	entry_ref ref = _GetCategoryRef(name);
+	BFile* catFile = new BFile(&ref, B_READ_ONLY);
```
Comment (pulkomandy) [inline src/db/QueryDBManager.cpp:360]: This is never deleted.

Use `BFile catFile(&ref, B_READ_ONLY);` and then `_FileToCategory(&catFile);

(the same pattern is present in other places in the code)

File: src/db/QueryDBManager.cpp [hunk 5]
```diff
@@ -0,0 +1,950 @@
...
+	BString name = BString(category->GetName());
+	file->WriteAttr("Category:Name", B_STRING_TYPE, 0, name.String(),
+					name.CountChars() + 1);
+	
+	BString id = BString(category->GetId());
+	file->WriteAttr("Calendar:ID", B_STRING_TYPE, 0, id.String(),
+						id.CountChars() + 1);
+
+	BString color = category->GetHexColor();
+	file->WriteAttr("Category:Color", B_STRING_TYPE, 0, color.String(),
```
Comment (pulkomandy) [inline src/db/QueryDBManager.cpp:625]: why not store the color as a B_COLOR_TYPE ? I think we have that.
Or a B_INT32_TYPE would work as well.

File: src/db/QueryDBManager.cpp [hunk 6]
```diff
@@ -0,0 +1,950 @@
...
+	BString place = event->GetPlace();
+	file->WriteAttr("Event:Place", B_STRING_TYPE, 0, place.String(),
+					place.CountChars() + 1);
+
+	BString desc = event->GetDescription();
+	file->WriteAttr("Event:Description", B_STRING_TYPE, 0, desc.String(),
+					desc.CountChars() + 1);
+
+	time_t updated = event->GetUpdated();
+	file->WriteAttr("Event:Updated", B_TIME_TYPE, 0, &updated,
```
Comment (pulkomandy) [inline src/db/QueryDBManager.cpp:678]: Not really your problem, but we should check the handling of B_TIME_TYPE. time_t is a different type on 32 and 64bit systems, and on 32bit it stops in the year 2038 (now too close for us to ignore the problem).

Is it possible to store the time as 64bit instead? This way you can have events for the end of the universe and other useful things to remember.

File: src/db/QueryDBManager.cpp [hunk 7]
```diff
@@ -0,0 +1,950 @@
...
+	query->Fetch();
+	entry_ref ref;
+
+	BFile* evFile;
+	Event* event;
+
+	while (query->GetNextRef(&ref) == B_OK) {
+		evFile = new BFile(&ref, B_WRITE_ONLY);
+		evFile->WriteAttr("Event:Category", B_STRING_TYPE, 0,
+						newCategory.String(), newCategory.CountChars() + 1);
```
Comment (pulkomandy) [inline src/db/QueryDBManager.cpp:711]: indentation is strange here

File: src/db/QueryDBManager.cpp [hunk 8]
```diff
@@ -0,0 +1,950 @@
...
+QueryDBManager::_CategoryStatusSwitch(status_t result)
+{
+	switch (result)
+	{
+		case B_BAD_VALUE:
+		{
+			BAlert* alert = new BAlert(B_TRANSLATE("Category file"),
+				B_TRANSLATE("Couldn't open category file because the path is not specified. "
+				"It usually means that the programmer made a mistake. "
+				"There is nothing you can do about it. Sorry."),
```
Comment (pulkomandy) [inline src/db/QueryDBManager.cpp:727]: There is! People should report bugs!

File: src/db/QueryDBManager.cpp [hunk 9]
```diff
@@ -0,0 +1,950 @@
...
+		fs_create_index(device, "Event:Category",	B_STRING_TYPE, 0);
+		fs_create_index(device, "Event:Place",		B_STRING_TYPE, 0);
+		fs_create_index(device, "Event:Description",B_STRING_TYPE, 0);
+		fs_create_index(device, "Event:Start",		B_INT32_TYPE, 0);
+		fs_create_index(device, "Event:End",		B_INT32_TYPE, 0);
+		fs_create_index(device, "Event:Updated",	B_INT32_TYPE, 0);
+		fs_create_index(device, "Event:Status",		B_STRING_TYPE, 0);
+		fs_create_index(device, "Calendar:ID",		B_STRING_TYPE, 0);
+		fs_create_index(device, "Category:Name",	B_STRING_TYPE, 0);
+		fs_create_index(device, "Category:Color",	B_STRING_TYPE, 0);
```
Comment (pulkomandy) [inline src/db/QueryDBManager.cpp:879]: Is an index by color really needed? Do you search events by their color?

File: src/db/QueryDBManager.h [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+						Category* newCategory);
+		Category*	GetCategory(const char* name);
+		Category*	EnsureCategory(const char* name);
+		BList*		GetAllCategories();
+		bool		RemoveCategory(Category* category);
+		bool		RemoveCategory(entry_ref categoryRef);
+
+private:
+
+	void			_Initialise();
```
Comment (pulkomandy) [inline src/db/QueryDBManager.h:56]: I think it's initialize with a z?

**Comment by humdingerb** [comment]
I tried this PR and it seems to work nicely, all my test events created with a SQL-using Calendar have been converted to single files. I like it!

With that change, it'd be nice to assign the filetype application/x-calendar-event with Calendar. A double-clicked event should open Calendar with that event. x-calendar-event needs a nice icon, ofc. :)

Nice PR! Got to leave the code review to PulkoMandy et. al...

**Comment by humdingerb** [comment]
The initial syncing of my Google Calendar worked (entered the code provided by Google etc.). Now, using the menu item to sync, crashes Calendar: https://oshi.at/hxHgWd/Calendar-5676-debug-14-12-2020-08-14-42.report

I haven't sync'ed the Google Calendar before, so I dunno if this crash is related to this PRs change, or there's always been an issue with syncing...

File: src/db/QueryDBManager.cpp (hunk 2 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp:101]: Those're the default UUIDs given to these categories by the SQL manager. IDs are only required for the Google sync code, and the default category was identified by ID-- but that's not the case now, so they don't really have to be these specific IDs anymore. A unique ID can be automatically made, so I might as well get rid of that part.

File: src/db/QueryDBManager.cpp (hunk 3 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp:258]: I think it's necessary to initialize it? The program crashes if I remove "new BQuery()"

Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp:258]: Strike that, it works perfectly. No particular reason to use a pointer, I'll switch.

File: src/db/QueryDBManager.cpp (hunk 7 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp]: Fix snuck into new commit, d0105dd.

File: src/db/QueryDBManager.cpp (hunk 4 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp]: Thanks-- got it with d0105dd.

File: src/db/QueryDBManager.cpp (hunk 3 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp]: Switched over in d0105dd

File: src/db/QueryDBManager.cpp (hunk 2 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp]: Got rid of it with c198d48

File: src/db/QueryDBManager.cpp (hunk 9 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp]: I added it to disallow the creation of events with an already used color... which I forgot to do, haha.

Yea, it doesn't make sense to make an index for something that's just used once, a loop over all categories probably makes more sense.

Removed the index and did this with 6a6f918

**Comment by JadedCtrl** [comment]
@humdinger
Yea, Google sync's been broken for a while. It shouldd be crashing the same way as before.

> x-calendar-event needs a nice icon, ofc. :)
Truue! I'm using some icons locally, but they're too god-awful to share haha.

File: src/db/QueryDBManager.cpp (hunk 6 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp:675]: I don't want anyone missing an End of The World party, but I don't know if that's possible without changes to support kit, yet.

File: src/db/QueryDBManager.h (hunk 10 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.h]: You're right, switched over (71b19ea)

File: src/db/QueryDBManager.cpp (hunk 5 above)
Comment (JadedCtrl) [reply src/db/QueryDBManager.cpp]: B_RGB_COLOR_TYPE is used internally anyway, so it doesn't make sense to store as a string haha. Moved to that with a600411

**Comment by humdingerb** [comment]
>> x-calendar-event needs a nice icon, ofc. :)
> Truue! I'm using some icons locally, but they're too god-awful to share. :P

I suppose slightly modified app icons could be used for event files. I assume only "event" files and "cancelled" (those are cancelled events?) need to have an icon, as "categories" files wouldn't be queried by the user.

Here is an event file with a green arrow pointing to the event on the calendar: https://oshi.at/Bmcxjo/calendar_event.hvif

Here a cancelled event with a red X: https://oshi.at/FRCooz/calendar_cancelled.hvif

It might be a good idea to put all used icons into an "artwork" folder in the repo.

**Comment by humdingerb** [comment]
Sorry to have distracted you in this PR, maybe split the icon/filetype work into a later PR, so this can be merged (if all code reviewers consent).

One more word on the icon/filetype though... :)
[edited: got a bit long, I opened issue #62 instead.]

**Comment by JadedCtrl** [comment]
That's a good point-- just reverted.

**Pull request merged**